build_runner_core/lib/src/asset/file_based.dart

```diff
@@ -61,81 +66,50 @@ class FileBasedAssetReader extends AssetReader implements AssetReaderState {
           'an input. Please ensure you have that package in your deps, or '
           'remove it from your input sets.');
     }
+    // TODO(davidmorgan): make this read via `filesystem`, currently it
+    // reads directly via `dart:io`.
     return glob
         .list(followLinks: true, root: packageNode.path)
         .where((e) => e is File && !path.basename(e.path).startsWith('._'))
         .cast<File>()
         .map((file) => _fileToAssetId(file, packageNode));
   }
-}
 
-/// Creates an [AssetId] for [file], which is a part of [packageNode].
-AssetId _fileToAssetId(File file, PackageNode packageNode) {
-  var filePath = path.normalize(file.absolute.path);
-  var relativePath = path.relative(filePath, from: packageNode.path);
-  return AssetId(packageNode.name, relativePath);
+  /// Creates an [AssetId] for [file], which is a part of [packageNode].
+  AssetId _fileToAssetId(File file, PackageNode packageNode) {
+    var filePath = path.normalize(file.absolute.path);
+    var relativePath = path.relative(filePath, from: packageNode.path);
+    return AssetId(packageNode.name, relativePath);
+  }
 }
 
 /// Basic [AssetWriter] which uses a [PackageGraph] to look up where to write
 /// files to disk.
 class FileBasedAssetWriter implements RunnerAssetWriter {
   final PackageGraph packageGraph;
+  final Filesystem filesystem;
 
-  FileBasedAssetWriter(this.packageGraph);
+  FileBasedAssetWriter(this.packageGraph)
+      : filesystem = IoFilesystem(assetPathProvider: packageGraph);
 
   @override
-  Future writeAsBytes(AssetId id, List<int> bytes) async {
-    var file = _fileFor(id, packageGraph);
-    await _descriptorPool.withResource(() async {
-      await file.create(recursive: true);
-      await file.writeAsBytes(bytes);
-    });
-  }
+  Future writeAsBytes(AssetId id, List<int> bytes) =>
+      filesystem.writeAsBytes(id, bytes);
 
   @override
   Future writeAsString(AssetId id, String contents,
-      {Encoding encoding = utf8}) async {
-    var file = _fileFor(id, packageGraph);
-    await _descriptorPool.withResource(() async {
-      await file.create(recursive: true);
-      await file.writeAsString(contents, encoding: encoding);
-    });
-  }
+          {Encoding encoding = utf8}) =>
+      filesystem.writeAsString(id, contents);
```

Comment:
Shouldn't you pass along the encoding as well?
Or not have options for enconding and always use utf8...

Reply:
Done, thanks.

This is from the generator-facing `AssetWriter` interface, so no choice about the signature :)